accounts/views.py

```diff
@@ -98,10 +100,177 @@ def signup(request):
+        friend = User.objects.get(username=request_username)
+    except User.DoesNotExist:
+        return HttpResponse("User not found.", status=404)
+    # check if user is friends with the other user
+    if not request.user.friends.filter(pk=friend.pk).exists():
+        return HttpResponse("You are not friends with this user.", status=400)
+    # remove user from each others friends list
+    with transaction.atomic():
+        request.user.friends.remove(friend)
+        friend.friends.remove(request.user)
+    return HttpResponse("Successfully removed user from friends.", status=200)
+
+
+def accept_friend_request(request):
+    """
+    This function handles when a user wants to accept a friend request they
+    have received.
+    """
+    # check request is POST and has username field
+    if request.method != "POST":
+        return redirect("landing_page:index")
+    try:
+        data = json.loads(request.body)
+        request_username = data["username"]
+    except (json.JSONDecodeError, KeyError):
+        return HttpResponse("Invalid JSON", status=400)
+    # check user is logged in and friend request exists
+    if not request.user.is_authenticated:
+        return redirect("accounts:login")
+    try:
+        friend_request = FriendRequest.objects.get(
+            sender=User.objects.get(username=request_username), receiver=request.user
+        )
+    except FriendRequest.DoesNotExist:
+        return HttpResponse("Friend request not found.", status=404)
+    # accept friend request by adding each user to the other's friends list,
+    # then delete the friend request
+    with transaction.atomic():
+        request.user.friends.add(friend_request.sender)
+        friend_request.sender.friends.add(request.user)
+        friend_request.delete()
+    return HttpResponse("Successfully accepted friend request.", status=200)
+
+
+def decline_friend_request(request):
+    """
+    This function handles when a user wants to decline a friend request they
+    have received.
+    """
+    # check request is POST and has username field
+    if request.method != "POST":
+        return redirect("landing_page:index")
+    try:
+        data = json.loads(request.body)
+        request_username = data["username"]
+    except (json.JSONDecodeError, KeyError):
+        return HttpResponse("Invalid JSON", status=400)
+    # check user is logged in and friend request exists
+    if not request.user.is_authenticated:
+        return redirect("accounts:login")
+    try:
+        friend_request = FriendRequest.objects.get(
+            sender=User.objects.get(username=request_username), receiver=request.user
+        )
+    except FriendRequest.DoesNotExist:
+        return HttpResponse("Friend request not found.", status=404)
+    # delete friend request
+    with transaction.atomic():
+        friend_request.delete()
+    return HttpResponse("Successfully rejected friend request.", status=200)
+
+
+def delete_friend_request(request):
```

Comment:
i feel like 'decline' and 'delete' are kind of an unnecessary distinction - can we just make it one? and then try the sender, if they aren't the sender, then try the receiver?

Reply:
Oh yeah didn't realize that, I think a boolean flag parameter can be sent in the POST to easily distinguish between the two, but is that a safe (security wise) change? I'll have to look into that more